### PR TITLE
check public parameters match

### PR DESCRIPTION
**Pull request by adecaro**
This PR adds a new check to the fungible integration test. Namely, the check verifies that the public parameters used by the FSC nodes matches those stored at the backend.

The API has been update to be able to fetch the public parameters from the backend. The API can be used also for debugging in case a mismatch appears.

Signed-off-by: Angelo De Caro <adc@zurich.ibm.com>

### CODE REVIEW

File: token/core/fabtoken/driver/driver.go [hunk 1]
```diff
@@ -93,7 +92,7 @@ func (d *Driver) NewTokenService(sp view2.ServiceProvider, publicParamsFetcher d
 		fabtoken.NewDeserializer(),
 		tmsConfig,
 	)
-	if err := service.PPM.ForceFetch(); err != nil {
+	if err := service.PPM.Update(); err != nil {
 		return nil, errors.WithMessage(err, "failed to fetch public parameters")
```
Comment (KElkhiyaoui) [inline token/core/fabtoken/driver/driver.go:96]: failed to fetch and update

File: token/core/fabtoken/loaders.go [hunk 2]
```diff
@@ -27,55 +29,38 @@ func (s *VaultTokenLoader) GetTokens(ids []*token.ID) ([]string, []*token.Token,
...
-			logger.Errorf("failed retrieving public params [%s]", err)
-			return nil, err
-		}
-	}
-
-	logger.Debugf("unmarshal public parameters")
-	pp := &PublicParams{}
-	pp.Label = s.PPLabel
-	err = pp.Deserialize(raw)
+// Fetch fetches the public parameters from the remote repository
```
Comment (KElkhiyaoui) [inline token/core/fabtoken/loaders.go:36]: what a remote repository?

File: token/core/fabtoken/loaders.go [hunk 3]
```diff
@@ -27,55 +29,38 @@ func (s *VaultTokenLoader) GetTokens(ids []*token.ID) ([]string, []*token.Token,
...
-	return pp, nil
+	logger.Debugf("fetched public parameters [%s]", hash.Hashable(raw).String())
+	return raw, nil
 }
 
-// ForceFetch returns the PublicParams associated with fabtoken
-// ForceFetch does not look into cache, unlike Load.
-func (s *VaultPublicParamsLoader) ForceFetch() (*PublicParams, error) {
-	logger.Debugf("force public parameters fetch")
+// FetchParams fetches the public parameters from the remote repository and unmarshal them
```
Comment (KElkhiyaoui) [inline token/core/fabtoken/loaders.go:48]: what remote repository?

File: token/core/fabtoken/ppm.go [hunk 4]
```diff
@@ -67,6 +72,18 @@ func (v *PublicParamsManager) ForceFetch() error {
 	return nil
 }
 
+// Fetch fetches the public parameters from the backed
```
Comment (KElkhiyaoui) [inline token/core/fabtoken/ppm.go:75]: the backend

File: token/core/fabtoken/service.go [hunk 5]
```diff
@@ -17,8 +17,8 @@ import (
 )
```
Comment (KElkhiyaoui) [inline token/core/fabtoken/service.go:18]: add comments to interface

File: token/core/zkatdlog/crypto/ppm/ppm.go [hunk 6]
```diff
@@ -9,15 +9,18 @@ package ppm
...
-
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view/services/flogging"
+	"github.com/hyperledger-labs/fabric-smart-client/platform/view/services/hash"
 	"github.com/hyperledger-labs/fabric-token-sdk/token/core/zkatdlog/crypto"
 	"github.com/hyperledger-labs/fabric-token-sdk/token/driver"
+	"github.com/pkg/errors"
 )
 
+var logger = flogging.MustGetLogger("token-sdk.driver.zkatdlog")
+
```
Comment (KElkhiyaoui) [inline token/core/zkatdlog/crypto/ppm/ppm.go:20]: add comments to interface

File: token/core/zkatdlog/crypto/ppm/ppm.go [hunk 7]
```diff
@@ -41,19 +44,24 @@ func (v *PublicParamsManager) PublicParameters() driver.PublicParameters {
...
 	defer v.mutex.Unlock()
 
 	if v.publicParamsLoader == nil {
 		return errors.New("public parameters loader not set")
 	}
 
-	pp, err := v.publicParamsLoader.ForceFetch()
+	pp, err := v.publicParamsLoader.FetchParams()
 	if err != nil {
 		return errors.WithMessagef(err, "failed force fetching public parameters")
```
Comment (KElkhiyaoui) [inline token/core/zkatdlog/crypto/ppm/ppm.go:66]: failed updating

File: token/core/zkatdlog/nogh/loaders.go [hunk 8]
```diff
@@ -94,58 +94,42 @@ func (s *VaultTokenLoader) LoadTokens(ids []*token3.ID) ([]string, []*token.Toke
...
-			logger.Errorf("failed retrieving public params [%s]", err)
-			return nil, err
-		}
-	}
-
-	logger.Debugf("unmarshal public parameters")
-	pp := &crypto.PublicParams{}
-	pp.Label = s.PPLabel
-	err = pp.Deserialize(raw)
+// Fetch fetches the public parameters from the backed
```
Comment (KElkhiyaoui) [inline token/core/zkatdlog/nogh/loaders.go:105]: from the backend

File: token/core/zkatdlog/nogh/loaders.go [hunk 9]
```diff
@@ -94,58 +94,42 @@ func (s *VaultTokenLoader) LoadTokens(ids []*token3.ID) ([]string, []*token.Toke
...
-	logger.Debugf("unmarshal public parameters done")
-	return pp, nil
+	logger.Debugf("fetched public parameters [%s]", hash.Hashable(raw).String())
+	return raw, nil
 }
 
-// ForceFetch retrieves the public parameters from the ledger
-func (s *VaultPublicParamsLoader) ForceFetch() (*crypto.PublicParams, error) {
-	logger.Debugf("force public parameters fetch")
+// FetchParams fetches the public parameters from the backed and unmarshals them
```
Comment (KElkhiyaoui) [inline token/core/zkatdlog/nogh/loaders.go:117]: from the backend

File: token/publicparams.go [hunk 10]
```diff
@@ -47,14 +47,14 @@ func (c *PublicParametersManager) MaxTokenValue() uint64 {
...
-	return c.ppm.PublicParameters().Bytes()
+// SerializePublicParameters returns the public parameters in their serialized form
+func (c *PublicParametersManager) SerializePublicParameters() ([]byte, error) {
+	return c.ppm.SerializePublicParameters()
 }
 
-// ForceFetch forcefully fetches the public parameters from the backend
-func (c *PublicParametersManager) ForceFetch() error {
-	return c.ppm.ForceFetch()
+// Update fetches the public parameters from the backend and update the local one accordingly
```
Comment (KElkhiyaoui) [inline token/publicparams.go:55]: the local ones

**Review by KElkhiyaoui — CHANGES_REQUESTED** [review]
please address comments.

**Comment by adecaro** [comment]
Hi @KElkhiyaoui , please, could you push the changes directly so that we can merge this? 😃

**Review by KElkhiyaoui — APPROVED** [review]
looks good to me

**Pull request merged**